File: solutions/csharp/connect/1/src/Connect.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Connect(string[] inputString)
{
	private readonly Board _board = Board.From(inputString);

	private static readonly Dictionary<Player, Func<Board, Player, bool>> _winningConditionByPlayer
		= new()
		{
			{ Player.White, IsVerticallyConnected },
			{ Player.Black, IsHorizontallyConnected }
		};

	public ConnectWinner Result()
	{
		foreach ((Player player, Func<Board, Player, bool> victoryCondition) in _winningConditionByPlayer)
		{
			if (victoryCondition(_board, player))
			{
				return player.To<ConnectWinner>();
			}
		}

		return ConnectWinner.None;
	}

	private static bool IsVerticallyConnected(Board board, Player player)
	{
		List<Cell> bottomEdge = board.GetBottomEdge();
		List<Cell> topEdge = board.GetTopEdge();
		return IsConnected(board, player, bottomEdge, topEdge);
	}

	private static bool IsHorizontallyConnected(Board board, Player player)
	{
		List<Cell> leftEdge = board.GetLeftEdge();
		List<Cell> rightEdge = board.GetRightEdge();
		return IsConnected(board, player, leftEdge, rightEdge);
	}

	private static bool IsConnected(
			Board board,
			Player player,
			List<Cell> startEdge,
			List<Cell> endEdge
		)
	{
		List<Cell> endCells = endEdge.BelongingTo(player);
		HashSet<HexCoordinates> visited = new();
		return startEdge.BelongingTo(player)
			.Any(startCell =>
				DepthFirstSearch(
					board,
					player,
					startCell.Coordinates,
					endCells,
					visited
				));
	}

	private static bool DepthFirstSearch(
			Board board,
			Player player,
			HexCoordinates currentCoordinates,
			List<Cell> endCells,
			HashSet<HexCoordinates> visited
		)
		=> endCells.Any(cell => cell.Coordinates.Equals(currentCoordinates))
			|| (visited.Add(currentCoordinates) && board.GetNeighbors(currentCoordinates)
				.Any(neighbor =>
					!visited.Contains(neighbor.Coordinates)
					&& neighbor.BelongsTo(player)
					&& DepthFirstSearch(board, player, neighbor.Coordinates, endCells, visited)
				));
}

File: solutions/java/connect/3/src/main/java/Connect.java
import java.util.*;
import java.util.function.BiFunction;
import java.util.stream.StreamSupport;

final class Connect {
    private final Board board;
    public Connect(String[] input) {
        this.board = Board.from(input);
    }

    private static final Map<Player, BiFunction<Board, Player, Boolean>> WINNING_CONDITIONS =
            Map.of(
                    Player.White, Connect::isVerticallyConnected,
                    Player.Black, Connect::isHorizontallyConnected
            );

    public Winner computeWinner() {
        for (Map.Entry<Player, BiFunction<Board, Player, Boolean>> entry : WINNING_CONDITIONS.entrySet()) {
            Player player = entry.getKey();
            BiFunction<Board, Player, Boolean> condition = entry.getValue();
            if (condition.apply(board, player)) {
                return player.winner;
            }
        }

        return Winner.NONE;
    }

    private static boolean isVerticallyConnected(Board board, Player player) {
        List<Cell> bottom = board.getBottomEdge();
        List<Cell> top = board.getTopEdge();
        return isConnected(board, player, bottom, top);
    }

    private static boolean isHorizontallyConnected(Board board, Player player) {
        List<Cell> left = board.getLeftEdge();
        List<Cell> right = board.getRightEdge();
        return isConnected(board, player, left, right);
    }

    private static boolean isConnected(
            Board board,
            Player player,
            List<Cell> startEdge,
            List<Cell> endEdge
    ) {
        List<Cell> endCells = belongingTo(endEdge, player);
        Set<HexCoordinates> visited = new HashSet<>();
        return belongingTo(startEdge, player)
                .stream()
                .anyMatch(startCell ->
                                  depthFirstSearch(
                                        board,
                                        player,
                                        startCell.coordinates(),
                                        endCells,
                                        visited
                                )
              );
    }

    private static boolean depthFirstSearch(
            Board board,
            Player player,
            HexCoordinates current,
            List<Cell> endCells,
            Set<HexCoordinates> visited
    ) {
        boolean isTarget
                = endCells.stream()
                          .anyMatch(cell -> cell.coordinates().equals(current));
        if (isTarget) {
            return true;
        }

        if (!visited.add(current)) {
            return false;
        }

        return board.getNeighbors(current)
                    .filter(neighbor -> !visited.contains(neighbor.coordinates()))
                    .filter(neighbor -> neighbor.belongsTo(player))
                    .anyMatch(neighbor ->
                                      depthFirstSearch(
                                              board,
                                              player,
                                              neighbor.coordinates(),
                                              endCells,
                                              visited
                                      )
                    );
    }

    static List<Cell> belongingTo(Iterable<Cell> cells, Player player) {
        BoardContent targetContent = player.boardContent;

        return StreamSupport.stream(cells.spliterator(), false)
                            .filter(cell -> cell.content() == targetContent)
                            .toList();
    }
}